Documentation around "query" parameter wrong?  Bad escaping/parsing?

- Version: 2.3.x
- Operating System: Linux

Hey team,

It looks like the documentation around the `query` parameter is incorrect.  It states:

```
Default value is "{\"query\": { \"match_all\": {} } }"
```

However, if I just copy and paste that in as the value for the query parameter, Elasticsearch throws a parsing exception. For example, with this input configuration:

```
input {
  # We read from the "old" index
  elasticsearch {
    hosts => ["http://localhost:9200"]
    index => "public_toilets"
    query => "{\"query\": { \"match_all\": {} } }"
    size => 500
    scroll => "5m"
    docinfo => true
  }
}
```

Logstash throws the following exception:

```
A plugin had an unrecoverable error. Will restart this plugin.
  Plugin: <LogStash::Inputs::Elasticsearch hosts=>["http://localhost:9200"], index=>"public_toilets", query=>"{ \\\"query\\\": { \\\"type\\\": { \\\"value\\\": \\\"toilets\\\" } } }", size=>500, scroll=>"5m", docinfo=>true, codec=><LogStash::Codecs::JSON charset=>"UTF-8">, scan=>true, docinfo_target=>"@metadata", docinfo_fields=>["_index", "_type", "_id"], ssl=>false>
  Error: [500] {"error":{"root_cause":[{"type":"json_parse_exception","reason":"Unexpected character ('\\' (code 92)): was expecting either valid name character (for unquoted name) or double-quote (for quoted) to start field name\n at [Source: [B@2fee7767; line: 1, column: 4]"}],"type":"search_phase_execution_exception","reason":"all shards failed","phase":"init_scan","grouped":true,"failed_shards":[{"shard":0,"index":"public_toilets","node":"FQEp7OpWTruaD_eOjiqLYg","reason":{"type":"json_parse_exception","reason":"Unexpected character ('\\' (code 92)): was expecting either valid name character (for unquoted name) or double-quote (for quoted) to start field name\n at [Source: [B@2fee7767; line: 1, column: 4]"}}]},"status":500} {:level=>:error}
```

And I see the following exception in the Elasticsearch logs:

```
es_1 | [2016-07-21 01:19:06,334][WARN ][rest.suppressed          ] path: /public_toilets/_search, params: {size=500, scroll=5m, index=public_toilets, search_type=scan}
es_1 | Failed to execute phase [init_scan], all shards failed; shardFailures {[FQEp7OpWTruaD_eOjiqLYg][public_toilets][0]: RemoteTransportException[[escluster1_es_1][172.18.0.2:9300][indices:data/read/search[phase/scan]]]; nested: SearchParseException[failed to parse search source [{\"query\": { \"match_all\": {} } }]]; nested: JsonParseException[Unexpected character ('\' (code 92)): was expecting either valid name character (for unquoted name) or double-quote (for quoted) to start field name
es_1 |  at [Source: [B@5f22c259; line: 1, column: 3]]; }
es_1 |  at org.elasticsearch.action.search.AbstractSearchAsyncAction.onFirstPhaseResult(AbstractSearchAsyncAction.java:206)
es_1 |  at org.elasticsearch.action.search.AbstractSearchAsyncAction$1.onFailure(AbstractSearchAsyncAction.java:152)
es_1 |  at org.elasticsearch.action.ActionListenerResponseHandler.handleException(ActionListenerResponseHandler.java:46)
es_1 |  at org.elasticsearch.transport.TransportService$DirectResponseChannel.processException(TransportService.java:855)
es_1 |  at org.elasticsearch.transport.TransportService$DirectResponseChannel.sendResponse(TransportService.java:833)
es_1 |  at org.elasticsearch.transport.DelegatingTransportChannel.sendResponse(DelegatingTransportChannel.java:68)
es_1 |  at org.elasticsearch.transport.RequestHandlerRegistry$TransportChannelWrapper.sendResponse(RequestHandlerRegistry.java:146)
es_1 |  at org.elasticsearch.shield.transport.ShieldServerTransportService$ProfileSecuredRequestHandler.messageReceived(ShieldServerTransportService.java:182)
es_1 |  at org.elasticsearch.transport.RequestHandlerRegistry.processMessageReceived(RequestHandlerRegistry.java:75)
es_1 |  at org.elasticsearch.transport.TransportService$4.doRun(TransportService.java:376)
es_1 |  at org.elasticsearch.common.util.concurrent.AbstractRunnable.run(AbstractRunnable.java:37)
es_1 |  at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1142)
es_1 |  at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:617)
es_1 |  at java.lang.Thread.run(Thread.java:745)
es_1 | Caused by: [Unexpected character ('\' (code 92)): was expecting either valid name character (for unquoted name) or double-quote (for quoted) to start field name
es_1 |  at [Source: [B@5f22c259; line: 1, column: 3]]; nested: JsonParseException[Unexpected character ('\' (code 92)): was expecting either valid name character (for unquoted name) or double-quote (for quoted) to start field name
es_1 |  at [Source: [B@5f22c259; line: 1, column: 3]];
es_1 |  at org.elasticsearch.ElasticsearchException.guessRootCauses(ElasticsearchException.java:386)
es_1 |  at org.elasticsearch.action.search.SearchPhaseExecutionException.guessRootCauses(SearchPhaseExecutionException.java:152)
es_1 |  at org.elasticsearch.action.search.SearchPhaseExecutionException.getCause(SearchPhaseExecutionException.java:99)
es_1 |  at java.lang.Throwable.printStackTrace(Throwable.java:665)
es_1 |  at java.lang.Throwable.printStackTrace(Throwable.java:721)
es_1 |  at org.apache.log4j.DefaultThrowableRenderer.render(DefaultThrowableRenderer.java:60)
es_1 |  at org.apache.log4j.spi.ThrowableInformation.getThrowableStrRep(ThrowableInformation.java:87)
es_1 |  at org.apache.log4j.spi.LoggingEvent.getThrowableStrRep(LoggingEvent.java:413)
es_1 |  at org.apache.log4j.WriterAppender.subAppend(WriterAppender.java:313)
es_1 |  at org.apache.log4j.WriterAppender.append(WriterAppender.java:162)
es_1 |  at org.apache.log4j.AppenderSkeleton.doAppend(AppenderSkeleton.java:251)
es_1 |  at org.apache.log4j.helpers.AppenderAttachableImpl.appendLoopOnAppenders(AppenderAttachableImpl.java:66)
es_1 |  at org.apache.log4j.Category.callAppenders(Category.java:206)
es_1 |  at org.apache.log4j.Category.forcedLog(Category.java:391)
es_1 |  at org.apache.log4j.Category.log(Category.java:856)
es_1 |  at org.elasticsearch.common.logging.log4j.Log4jESLogger.internalWarn(Log4jESLogger.java:135)
es_1 |  at org.elasticsearch.common.logging.support.AbstractESLogger.warn(AbstractESLogger.java:109)
es_1 |  at org.elasticsearch.rest.BytesRestResponse.convert(BytesRestResponse.java:134)
es_1 |  at org.elasticsearch.rest.BytesRestResponse.<init>(BytesRestResponse.java:96)
es_1 |  at org.elasticsearch.rest.BytesRestResponse.<init>(BytesRestResponse.java:87)
es_1 |  at org.elasticsearch.rest.action.support.RestActionListener.onFailure(RestActionListener.java:60)
es_1 |  at org.elasticsearch.action.support.TransportAction$1.onFailure(TransportAction.java:95)
es_1 |  at org.elasticsearch.shield.action.ShieldActionFilter$SigningListener.onFailure(ShieldActionFilter.java:203)
es_1 |  at org.elasticsearch.action.support.TransportAction$FilteredActionListener.onFailure(TransportAction.java:242)
es_1 |  at org.elasticsearch.action.search.AbstractSearchAsyncAction.raiseEarlyFailure(AbstractSearchAsyncAction.java:294)
es_1 |  ... 14 more
es_1 | Caused by: com.fasterxml.jackson.core.JsonParseException: Unexpected character ('\' (code 92)): was expecting either valid name character (for unquoted name) or double-quote (for quoted) to start field name
es_1 |  at [Source: [B@5f22c259; line: 1, column: 3]
es_1 |  at com.fasterxml.jackson.core.JsonParser._constructError(JsonParser.java:1581)
es_1 |  at com.fasterxml.jackson.core.base.ParserMinimalBase._reportError(ParserMinimalBase.java:533)
es_1 |  at com.fasterxml.jackson.core.base.ParserMinimalBase._reportUnexpectedChar(ParserMinimalBase.java:462)
es_1 |  at com.fasterxml.jackson.core.json.UTF8StreamJsonParser._handleOddName(UTF8StreamJsonParser.java:2012)
es_1 |  at com.fasterxml.jackson.core.json.UTF8StreamJsonParser._parseName(UTF8StreamJsonParser.java:1650)
es_1 |  at com.fasterxml.jackson.core.json.UTF8StreamJsonParser.nextToken(UTF8StreamJsonParser.java:740)
es_1 |  at org.elasticsearch.common.xcontent.json.JsonXContentParser.nextToken(JsonXContentParser.java:53)
es_1 |  at org.elasticsearch.search.SearchService.parseSource(SearchService.java:830)
es_1 |  at org.elasticsearch.search.SearchService.createContext(SearchService.java:654)
es_1 |  at org.elasticsearch.search.SearchService.createAndPutContext(SearchService.java:620)
es_1 |  at org.elasticsearch.search.SearchService.executeScan(SearchService.java:281)
es_1 |  at org.elasticsearch.search.action.SearchServiceTransportAction$SearchScanTransportHandler.messageReceived(SearchServiceTransportAction.java:425)
es_1 |  at org.elasticsearch.search.action.SearchServiceTransportAction$SearchScanTransportHandler.messageReceived(SearchServiceTransportAction.java:421)
es_1 |  at org.elasticsearch.transport.TransportRequestHandler.messageReceived(TransportRequestHandler.java:33)
es_1 |  at org.elasticsearch.shield.transport.ShieldServerTransportService$ProfileSecuredRequestHandler.messageReceived(ShieldServerTransportService.java:180)
es_1 |  ... 6 more
```

So it looks like somewhere, there is too much escaping happening?  

If I just use single quotes around the `query` parameter value and remove the backslashes, i.e.:

```
    query => '{"query": { "match_all": {} } }'
```

No exceptions are thrown.  

Is this a bug in the docs or a bug in the plugin?

--- a/docs/index.asciidoc
+++ b/docs/index.asciidoc
@@ -88,7 +88,8 @@ input plugins.
   * Value type is <<path,path>>
   * There is no default value for this setting.
 
-SSL Certificate Authority file in PEM encoded format, must also include any chain certificates as necessary 
+SSL Certificate Authority file in PEM encoded format, must also
+include any chain certificates as necessary.
 
 [id="plugins-{type}s-{plugin}-docinfo"]
 ===== `docinfo` 
@@ -111,7 +112,7 @@ Example
       elasticsearch {
         hosts => "es.production.mysite.org"
         index => "mydata-2018.09.*"
-        query => "*"
+        query => '{ "query": { "query_string": { "query": "*" } } }'
         size => 500
         scroll => "5m"
         docinfo => true
@@ -132,9 +133,11 @@ Example
   * Value type is <<array,array>>
   * Default value is `["_index", "_type", "_id"]`
 
-List of document metadata to move to the `docinfo_target` field
-To learn more about Elasticsearch metadata fields read
-http://www.elasticsearch.org/guide/en/elasticsearch/guide/current/_document_metadata.html
+If document metadata storage is requested by enabling the `docinfo`
+option, this option lists the metadata fields to save in the current
+event. See
+[Document Metadata](http://www.elasticsearch.org/guide/en/elasticsearch/guide/current/_document_metadata.html)
+in the Elasticsearch documentation for more information.
 
 [id="plugins-{type}s-{plugin}-docinfo_target"]
 ===== `docinfo_target` 
@@ -142,7 +145,9 @@ http://www.elasticsearch.org/guide/en/elasticsearch/guide/current/_document_meta
   * Value type is <<string,string>>
   * Default value is `"@metadata"`
 
-Where to move the Elasticsearch document information by default we use the @metadata field.
+If document metadata storage is requested by enabling the `docinfo`
+option, this option names the field under which to store the metadata
+fields as subfields.
 
 [id="plugins-{type}s-{plugin}-hosts"]
 ===== `hosts` 
@@ -150,9 +155,9 @@ Where to move the Elasticsearch document information by default we use the @meta
   * Value type is <<array,array>>
   * There is no default value for this setting.
 
-List of elasticsearch hosts to use for querying.
-each host can be either IP, HOST, IP:port or HOST:port
-port defaults to 9200
+List of one or more Elasticsearch hosts to use for querying. Each host
+can be either IP, HOST, IP:port, or HOST:port. The port defaults to
+9200.
 
 [id="plugins-{type}s-{plugin}-index"]
 ===== `index` 
@@ -168,17 +173,19 @@ The index or alias to search.
   * Value type is <<password,password>>
   * There is no default value for this setting.
 
-Basic Auth - password
+The password to use together with the username in the `user` option
+when authenticating to the Elasticsearch server. If set to an empty
+string authentication will be disabled.
 
 [id="plugins-{type}s-{plugin}-query"]
 ===== `query` 
 
   * Value type is <<string,string>>
-  * Default value is `"{ \"sort\": [ \"_doc\" ] }"`
+  * Default value is `'{ "sort": [ "_doc" ] }'`
 
-The query to be executed. Read the Elasticsearch query DSL documentation
-for more info
-https://www.elastic.co/guide/en/elasticsearch/reference/current/query-dsl.html
+The query to be executed. Read the
+[Elasticsearch query DSL documentation](https://www.elastic.co/guide/en/elasticsearch/reference/current/query-dsl.html)
+for more information.
 
 [id="plugins-{type}s-{plugin}-scroll"]
 ===== `scroll` 
@@ -204,7 +211,8 @@ This allows you to set the maximum number of hits returned per scroll.
   * Value type is <<boolean,boolean>>
   * Default value is `false`
 
-SSL
+If enabled, SSL will be used when communicating with the Elasticsearch
+server (i.e. HTTPS will be used instead of plain HTTP).
 
 [id="plugins-{type}s-{plugin}-user"]
 ===== `user` 
@@ -212,7 +220,9 @@ SSL
   * Value type is <<string,string>>
   * There is no default value for this setting.
 
-Basic Auth - username
+The username to use together with the password in the `password`
+option when authenticating to the Elasticsearch server. If set to an
+empty string authentication will be disabled.
 
 
 

